### PR TITLE
Arm thumb/thumb shift

### PR DESCRIPTION
**Pull request by vdf-git**
new function to determine bitstring for (most) thumb-2 immediate shifts (slightly different from ARM shifts) added to asm/arch/arm/armass.c.

### CODE REVIEW

File: libr/asm/arch/arm/armass.c [hunk 1]
```diff
@@ -330,6 +330,59 @@ static int getlist(char *op) {
...
+	char *space;
+	ut32 res = 0;
+	ut32 shift = false;
+	err = false;
+	ut32 argn;
+	ut32 i;
+	
+	strncpy (type, str, sizeof (type) - 1);
+	// XXX strcasecmp is probably unportable
+	if (!strcasecmp (type, shifts[5])) {
```
Comment (radare) [inline libr/asm/arch/arm/armass.c]: this is a gnuism, use r_str_case to uppercase type. and then use strcmp

File: libr/asm/arch/arm/armass.c [hunk 2]
```diff
@@ -330,6 +330,59 @@ static int getlist(char *op) {
...
+	char type[128];
+	char arg[128];
+	char *space;
+	ut32 res = 0;
+	ut32 shift = false;
+	err = false;
+	ut32 argn;
+	ut32 i;
+	
+	strncpy (type, str, sizeof (type) - 1);
```
Comment (radare) [inline libr/asm/arch/arm/armass.c]: just use strdup (str); and avoid the limits of the stack

File: libr/asm/arch/arm/armass.c (hunk 1 above)
Comment (radare) [reply libr/asm/arch/arm/armass.c]: well bsdism from 93 :P but it doesnt exist in windows

File: libr/asm/arch/arm/armass.c [hunk 3]
```diff
@@ -330,6 +330,59 @@ static int getlist(char *op) {
...
+		space = strchr (type, ' ');
+		if (!space) {
+			return 0;
+		}
+		*space = 0;
+		strncpy (arg, ++space, sizeof(arg) - 1);
+
+		for (i = 0; shifts[i]; i++) {
+			// XXX strcasecmp is probably unportable
+			if (!strcasecmp (type, shifts[i])) {
```
Comment (radare) [inline libr/asm/arch/arm/armass.c]: same here, no need to uppercase shifts, just upercase type and then strcmp

File: libr/asm/arch/arm/armass.c [hunk 4]
```diff
@@ -330,6 +330,60 @@ static int getlist(char *op) {
...
+	ut32 shift = false;
+	err = false;
+	ut32 argn;
+	ut32 i;
+	
+	type = strdup(str);
+
+	r_str_case(type,true);
+	
+	if (strcmp(type, shifts[5])) {
```
Comment (radare) [inline libr/asm/arch/arm/armass.c]: dont use negated logic, it makes code harder to read. use if (!strcmp (

also add a space before every (

File: libr/asm/arch/arm/armass.c [hunk 5]
```diff
@@ -330,6 +330,60 @@ static int getlist(char *op) {
...
+		// handle RRX alias case
+		res |= 3 << 20;
+		return res;
+	} else {
```
Comment (radare) [inline libr/asm/arch/arm/armass.c]: there's no need to else after a return statement. avoid unnecessary nestings

File: libr/asm/arch/arm/armass.c [hunk 6]
```diff
@@ -330,6 +330,60 @@ static int getlist(char *op) {
...
+				break;
+			}
+		}
+		if (!shift) {
+			err = true;
+			return 0;
+		}
+		res |= i << 20;
+		
+		argn = getnum(arg);
```
Comment (radare) [inline libr/asm/arch/arm/armass.c]: missing space

File: libr/asm/arch/arm/armass.c [hunk 7]
```diff
@@ -330,6 +330,60 @@ static int getlist(char *op) {
...
+	char *type;
+	char *arg;
+	char *space;
+	ut32 res = 0;
+	ut32 shift = false;
+	err = false;
+	ut32 argn;
+	ut32 i;
+	
+	type = strdup(str);
```
Comment (radare) [inline libr/asm/arch/arm/armass.c]: this leaks. you should free it before leaving this function

File: libr/asm/arch/arm/armass.c [hunk 8]
```diff
@@ -330,6 +330,60 @@ static int getlist(char *op) {
 	return list;
 }
 
+static ut32 thumb_getshift(const char *str) {
+	// only immediate shifts are ever used by thumb-2. Bit positions are different from ARM.
+	const char *shifts[] = {
+		"LSL", "LSR", "ASR", "ROR", 0, "RRX"
+	};
+	char *type;
```
Comment (radare) [inline libr/asm/arch/arm/armass.c]: dont define variables at the begining without value, we use c99 so you can declare and assign in the same place.

for example: char *type = strdup (str);

File: libr/asm/arch/arm/armass.c (hunk 4 above)
Comment (vdf-git) [reply libr/asm/arch/arm/armass.c]: it's just a bug - fixed in next commit

**Pull request merged**